File: packages/typescript/src/core/adapters/evm.ts
import { parseTransaction } from 'viem'
import { privateKeyToAccount } from 'viem/accounts'
import type { Wallet, Eip712Capable } from '../base.js'
import { SigningError } from '../errors.js'

export class EvmAdapter implements Wallet, Eip712Capable {
  private readonly account: ReturnType<typeof privateKeyToAccount>
  private readonly network: string

  constructor(privateKey: Uint8Array, network: string = 'eip155') {
    const hex = `0x${Buffer.from(privateKey).toString('hex')}` as `0x${string}`
    this.account = privateKeyToAccount(hex)
    this.network = network
  }

  async getAddress(): Promise<string> {
    return this.account.address
  }

  async signRaw(rawTx: Uint8Array): Promise<string> {
    try {
      const serialized = `0x${Buffer.from(rawTx).toString('hex')}` as `0x${string}`
      const parsed = parseTransaction(serialized)
      const { r: _r, s: _s, v: _v, yParity: _yParity, ...transaction } = parsed as Record<
        string,
        unknown
      >
      const sig = await this.account.signTransaction(
        transaction as Parameters<typeof this.account.signTransaction>[0],
      )
      return sig.slice(2)
    } catch (e) {
      throw new SigningError(`EVM sign_raw failed: ${e}`)
    }
  }

  async signTransaction(payload: Record<string, unknown>): Promise<string> {
    try {
      const sig = await this.account.signTransaction(
        payload as Parameters<typeof this.account.signTransaction>[0],
      )
      return sig.slice(2)
    } catch (e) {
      throw new SigningError(`EVM sign_transaction failed: ${e}`)
    }
  }

  async signMessage(msg: Uint8Array): Promise<string> {
    try {
      const sig = await this.account.signMessage({
        message: { raw: msg },
      })
      return sig.slice(2)
    } catch (e) {
      throw new SigningError(`EVM sign_message failed: ${e}`)
    }
  }

  async signTypedData(data: Record<string, unknown>): Promise<string> {
    try {
      const { domain, types, primaryType, message } = data as {
        domain: Record<string, unknown>
        types: Record<string, Array<{ name: string; type: string }>>
        primaryType: string
        message: Record<string, unknown>
      }

      // Remove EIP712Domain from types — viem adds it automatically
      const { EIP712Domain: _domain, ...messageTypes } = types

      const sig = await this.account.signTypedData({
        domain: domain as any,
        types: messageTypes as any,
        primaryType,
        message: message as any,
      })
      return sig.slice(2)
    } catch (e) {
      throw new SigningError(`EVM sign_typed_data failed: ${e}`)
    }
  }
}

File: packages/python/src/agent_wallet/core/adapters/evm.py
"""EvmSigner — Local EVM signing helper using eth-account."""

from __future__ import annotations

from agent_wallet.core.base import Eip712Capable, SignOptions, Wallet
from agent_wallet.core.errors import SigningError


class EvmSigner(Wallet, Eip712Capable):
    """EVM wallet using local ECDSA signing via eth-account."""

    def __init__(self, private_key: bytes, network: str = "eip155") -> None:
        from eth_account import Account

        self._account = Account.from_key(private_key)
        self._network = network

    async def get_address(self) -> str:
        return self._account.address

    async def sign_raw(self, raw_tx: bytes, options: SignOptions | None = None) -> str:
        try:
            signed = self._account.sign_transaction(raw_tx)
            return signed.raw_transaction.hex()
        except Exception as e:
            raise SigningError(f"EVM sign_raw failed: {e}") from e

    async def sign_transaction(self, payload: dict, options: SignOptions | None = None) -> str:
        try:
            signed = self._account.sign_transaction(payload)
            return signed.raw_transaction.hex()
        except Exception as e:
            raise SigningError(f"EVM sign_transaction failed: {e}") from e

    async def sign_message(self, msg: bytes, options: SignOptions | None = None) -> str:
        try:
            from eth_account.messages import encode_defunct

            signable = encode_defunct(primitive=msg)
            signed = self._account.sign_message(signable)
            return signed.signature.hex()
        except Exception as e:
            raise SigningError(f"EVM sign_message failed: {e}") from e

    async def sign_typed_data(self, data: dict, options: SignOptions | None = None) -> str:
        """Sign EIP-712 typed data.

        Expects data in full EIP-712 format:
        {
            "types": {"EIP712Domain": [...], "MyType": [...]},
            "primaryType": "MyType",
            "domain": {...},
            "message": {...}
        }
        """
        try:
            from eth_account.messages import encode_typed_data

            signable = encode_typed_data(full_message=data)
            signed = self._account.sign_message(signable)
            return signed.signature.hex()
        except Exception as e:
            raise SigningError(f"EVM sign_typed_data failed: {e}") from e
